File: elixir/lib/neutron/auth/jwt.ex
defmodule Neutron.Auth.JWT do
  @moduledoc """
  JWT token creation and verification using HMAC-SHA256.

  Provides a Guardian-style API for issuing and verifying JWT tokens,
  backed by the JOSE library for standards-compliant signing.

  ## Configuration

  Set `NEUTRON_SECRET_KEY` in your environment or pass the secret explicitly.

  ## Example

      # Sign a token
      {:ok, token} = Neutron.Auth.JWT.sign(%{user_id: 42, role: "admin"})

      # Verify a token
      {:ok, claims} = Neutron.Auth.JWT.verify(token)
      claims["user_id"]  # => 42

      # Sign with custom expiry (1 hour)
      {:ok, token} = Neutron.Auth.JWT.sign(%{user_id: 42}, ttl: 3600)
  """

  @default_ttl 86_400
  @algorithm "HS256"

  @type claims :: map()

  @doc """
  Signs a payload into a JWT token.

  ## Options

    * `:ttl` — time-to-live in seconds (default: 86400 = 24h)
    * `:secret` — signing secret (default: from NEUTRON_SECRET_KEY)
    * `:issuer` — token issuer (default: "neutron")
  """
  @spec sign(claims(), keyword()) :: {:ok, String.t()} | {:error, term()}
  def sign(payload, opts \\ []) do
    secret = Keyword.get(opts, :secret, get_secret())
    ttl = Keyword.get(opts, :ttl, @default_ttl)
    issuer = Keyword.get(opts, :issuer, "neutron")

    now = System.system_time(:second)

    claims =
      payload
      |> Map.merge(%{
        "iat" => now,
        "exp" => now + ttl,
        "iss" => issuer,
        "jti" => generate_jti()
      })

    jwk = JOSE.JWK.from_oct(secret)
    jws = JOSE.JWS.from_map(%{"alg" => @algorithm})

    {_, token} =
      JOSE.JWT.from_map(claims)
      |> JOSE.JWT.sign(jwk, jws)
      |> JOSE.JWS.compact()

    {:ok, token}
  rescue
    e -> {:error, Exception.message(e)}
  end

  @doc """
  Verifies a JWT token and returns the decoded claims.

  Checks signature, expiration, and issuer.

  ## Options

    * `:secret` — verification secret (default: from NEUTRON_SECRET_KEY)
    * `:issuer` — expected issuer (default: "neutron")
  """
  @spec verify(String.t(), keyword()) :: {:ok, claims()} | {:error, atom() | String.t()}
  def verify(token, opts \\ []) do
    secret = Keyword.get(opts, :secret, get_secret())
    issuer = Keyword.get(opts, :issuer, "neutron")

    jwk = JOSE.JWK.from_oct(secret)

    case JOSE.JWT.verify_strict(jwk, [@algorithm], token) do
      {true, %JOSE.JWT{fields: claims}, _jws} ->
        with :ok <- check_expiration(claims),
             :ok <- check_issuer(claims, issuer) do
          {:ok, claims}
        end

      {false, _, _} ->
        {:error, :invalid_signature}

      {:error, _reason} ->
        {:error, :invalid_token}
    end
  rescue
    _ -> {:error, :invalid_token}
  end

  @doc """
  Decodes a JWT token without verifying the signature.

  Useful for inspecting token contents. Do not trust the claims without verification.
  """
  @spec peek(String.t()) :: {:ok, claims()} | {:error, term()}
  def peek(token) do
    case JOSE.JWT.peek_payload(token) do
      %JOSE.JWT{fields: claims} -> {:ok, claims}
      _ -> {:error, :invalid_token}
    end
  rescue
    _ -> {:error, :invalid_token}
  end

  @doc """
  Returns the remaining TTL of a token in seconds, or :expired.
  """
  @spec ttl(String.t(), keyword()) :: {:ok, non_neg_integer()} | {:error, :expired | term()}
  def ttl(token, opts \\ []) do
    case verify(token, opts) do
      {:ok, claims} ->
        exp = Map.get(claims, "exp", 0)
        remaining = exp - System.system_time(:second)

        if remaining > 0 do
          {:ok, remaining}
        else
          {:error, :expired}
        end

      error ->
        error
    end
  end

  # --- Internal ---

  defp check_expiration(claims) do
    case Map.get(claims, "exp") do
      nil ->
        :ok

      exp when is_number(exp) ->
        if exp > System.system_time(:second) do
          :ok
        else
          {:error, :token_expired}
        end

      _ ->
        :ok
    end
  end

  defp check_issuer(claims, expected) do
    case Map.get(claims, "iss") do
      nil -> :ok
      ^expected -> :ok
      _ -> {:error, :invalid_issuer}
    end
  end

  defp get_secret do
    Neutron.Config.load().secret_key
  end

  defp generate_jti do
    :crypto.strong_rand_bytes(16) |> Base.url_encode64(padding: false)
  end
end

File: python/neutron/auth/jwt.py
"""JWT authentication — create, decode, middleware, and dependency.

Supports HS256 (symmetric), RS256 (RSA), and ES256 (ECDSA).
Uses PyJWT when available for RSA/ECDSA; falls back to stdlib HMAC for HS256.
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time
from typing import Any, Callable

from starlette.requests import Request
from starlette.responses import JSONResponse

from neutron.depends import _Depends
from neutron.error import AppError

_SUPPORTED_ALGORITHMS = {"HS256", "RS256", "ES256"}


def _b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


def _b64url_decode(s: str) -> bytes:
    padding = 4 - len(s) % 4
    if padding != 4:
        s += "=" * padding
    return base64.urlsafe_b64decode(s)


def create_token(
    payload: dict[str, Any],
    secret: str,
    *,
    expires_in: int = 3600,
    algorithm: str = "HS256",
) -> str:
    """Create a JWT token.

    Args:
        payload: Claims to encode.
        secret: Secret key (HS256) or PEM private key (RS256/ES256).
        expires_in: Token lifetime in seconds (default 1 hour).
        algorithm: HS256, RS256, or ES256.

    Returns:
        Encoded JWT string.
    """
    if algorithm not in _SUPPORTED_ALGORITHMS:
        raise ValueError(
            f"Unsupported algorithm: {algorithm}. "
            f"Supported: {', '.join(sorted(_SUPPORTED_ALGORITHMS))}"
        )

    now = int(time.time())
    claims = {**payload, "iat": now, "exp": now + expires_in}

    if algorithm in ("RS256", "ES256"):
        try:
            import jwt as pyjwt
        except ImportError:
            raise ImportError(
                f"{algorithm} requires PyJWT with cryptography: "
                "pip install PyJWT[crypto]"
            )
        return pyjwt.encode(claims, secret, algorithm=algorithm)

    # HS256 — stdlib implementation (no dependencies)
    header = {"alg": "HS256", "typ": "JWT"}
    segments = [
        _b64url_encode(json.dumps(header).encode()),
        _b64url_encode(json.dumps(claims).encode()),
    ]
    signing_input = f"{segments[0]}.{segments[1]}"
    sig = hmac.new(
        secret.encode(), signing_input.encode(), hashlib.sha256
    ).digest()
    segments.append(_b64url_encode(sig))
    return ".".join(segments)


def decode_token(
    token: str,
    secret: str,
    *,
    algorithms: list[str] | None = None,
    verify_exp: bool = True,
) -> dict[str, Any]:
    """Decode and verify a JWT token.

    Args:
        token: Encoded JWT string.
        secret: Secret key (HS256) or PEM public key (RS256/ES256).
        algorithms: Allowed algorithms (default: ["HS256"]).
        verify_exp: Whether to check expiration.

    Raises:
        AppError: If token is invalid or expired (401).
    """
    allowed = algorithms or ["HS256"]

    parts = token.split(".")
    if len(parts) != 3:
        raise AppError(
            status=401, code="invalid_token",
            title="Invalid Token", detail="Malformed JWT",
        )

    try:
        header = json.loads(_b64url_decode(parts[0]))
    except Exception:
        raise AppError(
            status=401, code="invalid_token",
            title="Invalid Token", detail="Failed to decode JWT header",
        )

    alg = header.get("alg", "HS256")
    if alg not in allowed:
        raise AppError(
            status=401, code="invalid_token",
            title="Invalid Token", detail=f"Unsupported algorithm: {alg}",
        )

    # RS256/ES256 — delegate to PyJWT
    if alg in ("RS256", "ES256"):
        try:
            import jwt as pyjwt
        except ImportError:
            raise AppError(
                status=500, code="config_error",
                title="Configuration Error",
                detail=f"{alg} requires PyJWT: pip install PyJWT[crypto]",
            )
        try:
            return pyjwt.decode(
                token, secret, algorithms=[alg],
                options={"verify_exp": verify_exp},
            )
        except pyjwt.ExpiredSignatureError:
            raise AppError(
                status=401, code="token_expired",
                title="Token Expired", detail="JWT has expired",
            )
        except pyjwt.InvalidTokenError as e:
            raise AppError(
                status=401, code="invalid_token",
                title="Invalid Token", detail=str(e),
            )

    # HS256 — stdlib verification
    try:
        payload = json.loads(_b64url_decode(parts[1]))
        signature = _b64url_decode(parts[2])
    except Exception:
        raise AppError(
            status=401, code="invalid_token",
            title="Invalid Token", detail="Failed to decode JWT",
        )

    signing_input = f"{parts[0]}.{parts[1]}"
    expected = hmac.new(
        secret.encode(), signing_input.encode(), hashlib.sha256
    ).digest()
    if not hmac.compare_digest(signature, expected):
        raise AppError(
            status=401, code="invalid_token",
            title="Invalid Token", detail="Invalid signature",
        )

    if verify_exp and "exp" in payload:
        if time.time() > payload["exp"]:
            raise AppError(
                status=401, code="token_expired",
                title="Token Expired", detail="JWT has expired",
            )

    return payload


class JWTMiddleware:
    """ASGI middleware that validates JWT tokens from the Authorization header.

    Sets ``request.state.user`` with the decoded payload on success.
    Requests without a token pass through (use ``get_current_user`` to enforce).
    """

    def __init__(
        self,
        secret: str,
        *,
        exclude_paths: list[str] | None = None,
        algorithms: list[str] | None = None,
    ) -> None:
        if len(secret) < 32:
            raise ValueError(
                f"JWT secret must be at least 32 characters for HS256 security. Got {len(secret)}."
            )
        self.secret = secret
        self.exclude_paths = set(exclude_paths or ["/health", "/docs", "/openapi.json"])
        self.algorithms = algorithms

    def as_starlette_middleware(self) -> Any:
        from starlette.middleware import Middleware

        return Middleware(
            _JWTMiddlewareImpl,
            secret=self.secret,
            exclude_paths=self.exclude_paths,
            algorithms=self.algorithms,
        )


class _JWTMiddlewareImpl:
    def __init__(self, app: Any, secret: str, exclude_paths: set, algorithms: list | None) -> None:
        self.app = app
        self.secret = secret
        self.exclude_paths = exclude_paths
        self.algorithms = algorithms

    async def __call__(self, scope: dict, receive: Any, send: Any) -> None:
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        path = scope.get("path", "")
        if path in self.exclude_paths:
            await self.app(scope, receive, send)
            return

        request = Request(scope, receive)
        auth_header = request.headers.get("authorization", "")

        if auth_header.startswith("Bearer "):
            token = auth_header[7:]
            try:
                payload = decode_token(token, self.secret, algorithms=self.algorithms)
                scope.setdefault("state", {})["user"] = payload
            except AppError:
                # Invalid token — let it through, get_current_user will raise
                scope.setdefault("state", {})["user"] = None
        else:
            scope.setdefault("state", {})["user"] = None

        await self.app(scope, receive, send)


def get_current_user(request: Request) -> dict[str, Any]:
    """Dependency that extracts the current user from JWT.

    Usage::

        @router.get("/me")
        async def me(user: dict = Depends(get_current_user)) -> dict: ...
    """
    user = getattr(request.state, "user", None)
    if user is None:
        raise AppError(
            status=401,
            code="unauthorized",
            title="Unauthorized",
            detail="Authentication required",
        )
    return user
